### DISCUSSION TITLE
Share UART for reading data from two sensors

### DISCUSSION THREAD

**Discussion by dst10** [micropython/micropython#11956] [ESP32]
Hi,
I'm trying to read data from 4 sensors (via modbus RTU protocol) using an esp32. As this board has only 3 UARTs (UART0,UART1,UART2), I wonder if it would be a problem or a bad practice in the long term to share a port for two sensors and continously (once a minute) overwrite this UART port and reinitialize it with the communication parameters (pins,baudrate,parity,etc.) of the other sensor.

(I don't need to read data from sensors simultanously, it could be sequently)

Thanks!

**Comment by sosi-deadeye** [comment]
Modbus uses mostly RS485. Then you require a `RS485` to `UART 3.3V` converter. Each Modbus slave has a `unitid` to address the slave. With the converter you need only one free UART and a library, which handles the Modbus Protocol.

**Comment by dst10** [comment]
Thank you @sosi-deadeye . I think I didn't explain myself clear enough. I have 4 sensors and then 4 RS485 to TTL converters. Each of these converter has A and B terminal wired to a pair of esp32 pins. As esp32 has only 3 UART ports available, my doubt is if there will be any issue if a constantly do this for reading data from sensors 3 and 4:

```
uart = UART(id=1,tx="tx_pin_sensor3",rx="rx_pin_sensor3",...)
uart.read() #Read data from sensor 3
uart = UART(id=1,tx="tx_pin_sensor4",rx="rx_pin_sensor4",...) #Overwrite UART1
uart.read() #Read data from sensor 4
```

**Comment by peterhinch** [comment]
You'd need to ensure that the unused TX pin was set high immediately after moving the UART to another pin. So long as the sensors only transmit when requested I think it might work. If they send unsolicited messages it would be a can of worms.

**Comment by karfas** [comment]
Unless you have very tight timing requirements or your sensors aren't addressable, I don't see the point of using multiple UARTs at all. 
You can connect everything to a single RS485 **bus** and sequentially poll your sensors, so saving three of the four RS485 converters and a lot of cabling as well.

**Comment by dst10** [comment]
Thank you @peterhinch and @karfas . It makes perfect sense. @karfas you mention a really interesting solution of using a single RS485 to TTL converter, no matter the number of sensors if you read data sequentially. I can't figure out how if you're using a single converter (and as I understand if you have 4 sensors you connect 4 wires to A pin of the converter and 4 wires to B pin of the converter) you can identify which sensor to communicate with

**Reply by karfas** [reply]
Just use Google and find some RS485 tutorial (including cables). The star topology you describe will work only for short distances, whereas a RS485 **bus** might be a few km long (for low data rates like 9600 Bd).
It will only work with **adressable** sensors in half-duplex mode with the ESP32 as master.
